### ISSUE TITLE
iproute2 version 4.15.0-1 stops FireQOS from classifying packets

### ISSUE THREAD

**Issue by seii** [firehol/firehol#317]
I recently performed an upgrade on an Arch Linux ARM system and went from version `4.14.1-2` of the `iproute2` package to version `4.15.0-1`. Once I did so, running `fireqos status <name>` would display all the packet categories from `fireqos.conf`, but no traffic would actually classify into the categories and all rows would permanently display `-` instead of numbers. All packets would show under the `TOTAL` column, and would often have the "delayed" purple tag applied to that column.  

This behavior was observed across Linux kernel versions `4.14.3-1`, `4.15.6-1`, and `4.15.7-1`. (I'm guessing this means it's not a kernel problem.)

Using the configuration file below (real IPs replaced with masking):
```
FIREQOS_CONNMARK_RESTORE="act_connmark"

DEVICE=extern
INPUT_SPEED=3217kbps
OUTPUT_SPEED=570kbps
LINKTYPE="adsl remote bridged-llc"

INPUT_MIN_SPEED=263kbit
OUTPUT_MIN_SPEED=48kbit

WORK_LAPTOP="X.X.X.X"
VOIP_PHONE="Y.Y.Y.Y"

WORK_MARK="connmark A"
VOIP_MARK="connmark B"

VOIP_RATE="90kbps"

interface $DEVICE receive input rate $INPUT_SPEED $LINKTYPE
   class voip commit $VOIP_RATE prio 2
      #match dst $VOIP_PHONE
      match $VOIP_MARK
   class work commit 40% prio 3
      #match dst $WORK_LAPTOP
      match $WORK_MARK
   class default

interface $DEVICE transmit output rate $OUTPUT_SPEED $LINKTYPE
   class voip commit $VOIP_RATE prio 2
      #match src $VOIP_PHONE
      match $VOIP_MARK
      match4 udp src "${VOIP_PHONE}"
   class work commit 55% prio 3
      #match src $WORK_LAPTOP
      match $WORK_MARK
   class default
```

With iproute2 v4.14.1-2:
```
FIREQOS START
-------------------------------------

: interface extern receive input rate 3217kbps adsl remote bridged-llc (extern-ifb, 26353kbit, mtu 1500, quantum 1500, minrate 263kbit)
:       class voip commit 90kbps prio 2 (1:11, 737/26353kbit, prio 2)
:       class work commit 40% prio 3 (1:12, 10541/26353kbit, prio 3)
:       class default (1:8000, 263/26353kbit, prio 4)
:       committed rate 11541kbit (43%), the remaining 14811kbit will be spare bandwidth.

: interface extern transmit output rate 570kbps adsl remote bridged-llc (extern, 4669kbit, mtu 1500, quantum 1500, minrate 46kbit)
:       class voip commit 90kbps prio 2 (1:11, 737/4669kbit, prio 2)
:       class work commit 55% prio 3 (1:12, 2567/4669kbit, prio 3)
:       class default (1:8000, 46/4669kbit, prio 4)
:       committed rate 3350kbit (71%), the remaining 1318kbit will be spare bandwidth.


TC QDISC SHOW DEV EXTERN
---------------------------

qdisc htb 1: root refcnt 2 r2q 3 default 8000 direct_packets_stat 0 direct_qlen 1000
qdisc fq_codel 12: parent 1:12 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn
qdisc fq_codel 11: parent 1:11 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn
qdisc fq_codel 13: parent 1:8000 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn
qdisc ingress ffff: parent ffff:fff1 ----------------


TC CLASS SHOW DEV EXTERN
--------------------------

class htb 1:11 parent 1:1 leaf 11: prio 2 rate 737Kbit ceil 4669Kbit burst 1599b cburst 1599b
class htb 1:8000 parent 1:1 leaf 13: prio 4 rate 46Kbit ceil 4669Kbit burst 1599b cburst 1599b
class htb 1:1 root rate 4669Kbit ceil 4669Kbit burst 1599b cburst 1599b
class htb 1:12 parent 1:1 leaf 12: prio 3 rate 2567Kbit ceil 4669Kbit burst 1599b cburst 1599b
class fq_codel 11:359 parent 11:
class fq_codel 13:155 parent 13:
class fq_codel 13:1a2 parent 13:
class fq_codel 13:279 parent 13:
class fq_codel 13:3c2 parent 13:
class fq_codel 11:359 parent 11:
class fq_codel 13:155 parent 13:
class fq_codel 13:1a2 parent 13:
class fq_codel 13:279 parent 13:
class fq_codel 13:3c2 parent 13:


TC FILTER SHOW DEV EXTERN
----------------------------

filter parent 1: protocol ip pref 10 u32 chain 0
filter parent 1: protocol ip pref 10 u32 chain 0 fh 800: ht divisor 1
filter parent 1: protocol ip pref 10 u32 chain 0 fh 800::800 order 2048 key ht 800 bkt 0 flowid 1:11 not_in_hw
  mark 0x0004 0x003f (success 3)
filter parent 1: protocol ip pref 20 u32 chain 0
filter parent 1: protocol ip pref 20 u32 chain 0 fh 801: ht divisor 1
filter parent 1: protocol ip pref 20 u32 chain 0 fh 801::800 order 2048 key ht 801 bkt 0 flowid 1:11 not_in_hw
  match 00110000/00ff0000 at 8
  match c0a8002c/ffffffff at 12
filter parent 1: protocol ip pref 30 u32 chain 0
filter parent 1: protocol ip pref 30 u32 chain 0 fh 802: ht divisor 1
filter parent 1: protocol ip pref 30 u32 chain 0 fh 802::800 order 2048 key ht 802 bkt 0 flowid 1:12 not_in_hw
  mark 0x0003 0x003f (success 0)
```

With iproute2 version 4.15.0-1:
```
FIREQOS START
-------------------------------------

: interface extern receive input rate 3217kbps adsl remote bridged-llc (extern-ifb, 26353kbit, mtu 1500, quantum 1500, minrate 263kbit)
:       class voip commit 90kbps prio 2 (1:11, 737/26353kbit, prio 2)
:       class work commit 40% prio 3 (1:12, 10541/26353kbit, prio 3)
:       class default (1:8000, 263/26353kbit, prio 4)
:       committed rate 11541kbit (43%), the remaining 14811kbit will be spare bandwidth.

: interface extern transmit output rate 570kbps adsl remote bridged-llc (extern, 4669kbit, mtu 1500, quantum 1500, minrate 46kbit)
:       class voip commit 90kbps prio 2 (1:11, 737/4669kbit, prio 2)
:       class work commit 55% prio 3 (1:12, 2567/4669kbit, prio 3)
:       class default (1:8000, 46/4669kbit, prio 4)
:       committed rate 3350kbit (71%), the remaining 1318kbit will be spare bandwidth.


TC QDISC SHOW DEV EXTERN
---------------------------

qdisc htb 1: root refcnt 2 r2q 3 default 32768 direct_packets_stat 1 direct_qlen 1000
qdisc fq_codel 12: parent 1:12 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn
qdisc fq_codel 11: parent 1:11 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn
qdisc fq_codel 13: parent 1:8000 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn
qdisc ingress ffff: parent ffff:fff1 ----------------


TC CLASS SHOW DEV EXTERN
--------------------------

class htb 1:11 parent 1:1 leaf 11: prio rate 737Kbit ceil 4669Kbit burst 1599b cburst 1599b
class htb 1:8000 parent 1:1 leaf 13: prio rate 46Kbit ceil 4669Kbit burst 1599b cburst 1599b
class htb 1:1 root rate 4669Kbit ceil 4669Kbit burst 1599b cburst 1599b
class htb 1:12 parent 1:1 leaf 12: prio rate 2567Kbit ceil 4669Kbit burst 1599b cburst 1599b
class fq_codel 13:df parent 13:
class fq_codel 13:fd parent 13:
class fq_codel 13:3dd parent 13:
class fq_codel 13:df parent 13:
class fq_codel 13:fd parent 13:
class fq_codel 13:3dd parent 13:

TC FILTER SHOW DEV EXTERN
----------------------------

filter parent 1: protocol ip pref 10 u32 chain 0
filter parent 1: protocol ip pref 10 u32 chain 0 fh 800: ht divisor 1
filter parent 1: protocol ip pref 10 u32 chain 0 fh 800::800 order 2048 key ht 800 bkt 0 flowid 1:11 not_in_hw
  mark 0x0004 0x003f (success 0)
filter parent 1: protocol ip pref 20 u32 chain 0
filter parent 1: protocol ip pref 20 u32 chain 0 fh 801: ht divisor 1
filter parent 1: protocol ip pref 20 u32 chain 0 fh 801::800 order 2048 key ht 801 bkt 0 flowid 1:11 not_in_hw
  match 00110000/00ff0000 at 8
  match c0a8002c/ffffffff at 12
filter parent 1: protocol ip pref 30 u32 chain 0
filter parent 1: protocol ip pref 30 u32 chain 0 fh 802: ht divisor 1
filter parent 1: protocol ip pref 30 u32 chain 0 fh 802::800 order 2048 key ht 802 bkt 0 flowid 1:12 not_in_hw
  mark 0x0003 0x003f (success 0)
```

**Comment by ktsaou** [comment]
Hi, I confirm that fireqos status shows wrong info.
However the classification works - if you install netdata you will see the traffic classified properly.

I think iproute changed something that broken the parsing made by the fireqos.

**Comment by seii** [comment]
@ktsaou Okay, good to know. My netdata instance has never picked up the names set in FireQOS for some reason, so I couldn't check that way.

I don't have hard traffic data to back this up, but the newer `iproute2` package also "felt" like QoS wasn't being performed within my network. It might be worth checking if this is a parsing issue or if `tc` is actually broken in some way for this version. (You may have already looked, sorry if you have.)

**Comment by ktsaou** [comment]
I see there are a few issues with the new output (below copied from your message):

```
class htb 1:11 parent 1:1 leaf 11: prio rate 737Kbit ceil 4669Kbit burst 1599b cburst 1599b
class htb 1:8000 parent 1:1 leaf 13: prio rate 46Kbit ceil 4669Kbit burst 1599b cburst 1599b
class htb 1:1 root rate 4669Kbit ceil 4669Kbit burst 1599b cburst 1599b
class htb 1:12 parent 1:1 leaf 12: prio rate 2567Kbit ceil 4669Kbit burst 1599b cburst 1599b
```

`prio` does not have a value.

---

Also I see they removed the `rate`

---

Let me find a fix...

**Comment by ktsaou** [comment]
ok, the fix is in PR #318 
Keep in mind though that the missing `prio` is a bug of `iproute`...

**Comment by ktsaou** [comment]
merged it.

**Comment by ktsaou** [comment]
> My netdata instance has never picked up the names set in FireQOS for some reason, so I couldn't check that way.

This is strange. Can you help me trace it?

**Comment by seii** [comment]
Sure, what would you need for best troubleshooting? I'm using netdata version `1.9.0-1`. I do get the "Quality of Service" header in netdata, as well as the colors to show different classes, but the names of the classes are (using the previous config example):

```
15: 
19: 
11: 
13: 
```

instead of corresponding to the names in `fireqos.conf`.

**Comment by ktsaou** [comment]
Please run:

```
/usr/libexec/netdata/plugins.d/tc-qos-helper.sh 10
```

At the end it should say something like:

```
SETDEVICENAME world-in
SETCLASSNAME 12: dsl/voip
SETCLASSNAME 13: dsl/interactive
SETCLASSNAME 14: dsl/vpns
SETCLASSNAME 15: dsl/servers
SETCLASSNAME 16: dsl/synacks
SETCLASSNAME 17: dsl/surfing
SETCLASSNAME 18: dsl/default
SETCLASSNAME 19: dsl/background
SETCLASSNAME 20: default
SETDEVICEGROUP eth2
END eth2-ifb
WORKTIME 0
```

If it does not, then somehow it cannot find `fireqos`.

The logic is in the script:

```sh
fireqos_run_dir="/var/run/fireqos"

if [ ! -d "${fireqos_run_dir}" ]
    then

    # the fireqos executable - we will use it to find its config
    fireqos="$(which fireqos 2>/dev/null || command -v fireqos 2>/dev/null)"

    if [ ! -z "${fireqos}" ]
        then

        fireqos_exec_dir="$(realdir ${fireqos})"

        if [ ! -z "${fireqos_exec_dir}" -a "${fireqos_exec_dir}" != "." -a -f "${fireqos_exec_dir}/install.config" ]
            then

            LOCALSTATEDIR=
            source "${fireqos_exec_dir}/install.config"

            if [ -d "${LOCALSTATEDIR}/run/fireqos" ]
                then
                fireqos_run_dir="${LOCALSTATEDIR}/run/fireqos"
            fi
        fi
    fi
fi
```

So, if `/var/run/fireqos` does not exist, it attempts to find the locate of fireqos, it reads `install.config` from that directory to find `${LOCALSTATEDIR}`.

Something in the above, does not work for your setup. This could happen if you have installed fireqos on an unusual location (not in the system path).

**Comment by seii** [comment]
My particular setup does include (for example) the following when I run that command:

```
SETDEVICENAME receive
SETCLASSNAME 11: voip
SETCLASSNAME 12: work
SETCLASSNAME 13: default
SETDEVICEGROUP extern
```

Those are correct, those entries just don't happen to show up in the web interface. I'm not sure why, I've tried refreshing the web page among other things.

**Comment by ktsaou** [comment]
hm... have you tried restarting netdata?

**Comment by seii** [comment]
Yes, I have restarted it. Just in case, I restarted the service again and refreshed the browser. Unfortunately, it still shows only the class numbers.

**Comment by ktsaou** [comment]
hm...


Please fetch `http://your.netdata.ip:19999/api/v1/charts` and paste below the section `tc.INTERFACE`, like this:

```json
		"tc.br0": 		{
			"id": "tc.br0",
			"name": "tc.world_out",
			"type": "tc",
			"family": "br0",
			"context": "tc.qos",
			"title": "Class Usage (tc.world_out)",
			"priority": 7000,
			"plugin": "tc",
			"module": "",
			"enabled": true,
			"units": "kilobits/s",
			"data_url": "/api/v1/data?chart=tc.world_out",
			"chart_type": "stacked",
			"duration": 3996,
			"first_entry": 1520278681,
			"last_entry": 1520282677,
			"update_every": 1,
			"dimensions": {
				"15:": { "name": "dsl/servers" },
				"19:": { "name": "dsl/background" },
				"17:": { "name": "dsl/surfing" },
				"14:": { "name": "dsl/vpns" },
				"12:": { "name": "dsl/voip" },
				"16:": { "name": "dsl/synacks" },
				"18:": { "name": "dsl/default" },
				"20:": { "name": "default" },
				"13:": { "name": "dsl/interactive" }
			},
			"green": null,
			"red": null,
			"alarms": {

			}
		},
```

**Comment by seii** [comment]
For the IFB interface that FireQOS creates (the normal interface is basically the same):
```
"tc.extern-ifb": 		{
			"id": "tc.extern-ifb",
			"name": "tc.extern_ifb",
			"type": "tc",
			"family": "extern-ifb",
			"context": "tc.qos",
			"title": "Class Usage (tc.extern_ifb)",
			"priority": 7000,
			"plugin": "tc",
			"module": "",
			"enabled": true,
			"units": "kilobits/s",
			"data_url": "/api/v1/data?chart=tc.extern_ifb",
			"chart_type": "stacked",
			"duration": 3996,
			"first_entry": 1520280465,
			"last_entry": 1520284461,
			"update_every": 1,
			"dimensions": {
				"15:": { "name": "15:" },
				"19:": { "name": "19:" },
				"17:": { "name": "17:" },
				"14:": { "name": "14:" },
				"12:": { "name": "12:" },
				"16:": { "name": "16:" },
				"18:": { "name": "18:" },
				"11:": { "name": "11:" },
				"13:": { "name": "13:" }
			},
			"green": null,
			"red": null,
			"alarms": {

			}
```

**Comment by ktsaou** [comment]
hm... something else is happening.

I see in the logs above that you have:

```
SETDEVICENAME receive
SETCLASSNAME 11: voip
SETCLASSNAME 12: work
SETCLASSNAME 13: default
SETDEVICEGROUP extern
```

Could you please post the above from `BEGIN` to `END`, and also find the same interface on `/api/v1/charts`?

I think that somehow the device names are messed.

**Comment by seii** [comment]
If it helps, my network has "extern" as the interface, and "extern-ifb" is the IFB device that FireQOS is creating to handle ingress QoS on that interface.

tc-qos-helper results in full:
```
BEGIN extern
qdisc htb 1: root refcnt 2 r2q 3 default 8000 direct_packets_stat 0 direct_qlen 1000
 Sent 126988 bytes 712 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
qdisc fq_codel 12: parent 1:12 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 0 bytes 0 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 11: parent 1:11 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 8533 bytes 14 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 1113 drop_overlimit 0 new_flow_count 1 ecn_mark 0
  new_flows_len 0 old_flows_len 1
qdisc fq_codel 13: parent 1:8000 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 118455 bytes 698 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 1166 drop_overlimit 0 new_flow_count 196 ecn_mark 0
  new_flows_len 1 old_flows_len 2
qdisc ingress ffff: parent ffff:fff1 ---------------- 
 Sent 254184 bytes 698 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
SETDEVICENAME transmit
SETCLASSNAME 11: voip
SETCLASSNAME 12: work
SETCLASSNAME 13: default
SETDEVICEGROUP extern
END extern
BEGIN extern-ifb
qdisc htb 1: root refcnt 2 r2q 21 default 8000 direct_packets_stat 0 direct_qlen 32
 Sent 323194 bytes 698 pkt (dropped 0, overlimits 1 requeues 0) 
 backlog 0b 0p requeues 0 
qdisc fq_codel 12: parent 1:12 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 0 bytes 0 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 11: parent 1:11 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 4823 bytes 7 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 689 drop_overlimit 0 new_flow_count 1 ecn_mark 0
  new_flows_len 0 old_flows_len 1
qdisc fq_codel 13: parent 1:8000 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 318371 bytes 691 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 1696 drop_overlimit 0 new_flow_count 186 ecn_mark 0
  new_flows_len 1 old_flows_len 2
SETDEVICENAME receive
SETCLASSNAME 11: voip
SETCLASSNAME 12: work
SETCLASSNAME 13: default
SETDEVICEGROUP extern
END extern-ifb
BEGIN eth0
qdisc mq 0: root 
 Sent 1698171721 bytes 2677613 pkt (dropped 0, overlimits 0 requeues 6) 
 backlog 0b 0p requeues 6 
qdisc fq_codel 0: parent :8 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 2794653 bytes 9577 pkt (dropped 0, overlimits 0 requeues 1) 
 backlog 0b 0p requeues 1 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent :7 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 310535611 bytes 501834 pkt (dropped 0, overlimits 0 requeues 2) 
 backlog 0b 0p requeues 2 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent :6 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 720109 bytes 5567 pkt (dropped 0, overlimits 0 requeues 1) 
 backlog 0b 0p requeues 1 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent :5 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 559196202 bytes 940705 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent :4 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 1225494 bytes 6520 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent :3 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 367688508 bytes 556244 pkt (dropped 0, overlimits 0 requeues 1) 
 backlog 0b 0p requeues 1 
  maxpacket 387 drop_overlimit 0 new_flow_count 1 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent :2 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 1243187 bytes 7261 pkt (dropped 0, overlimits 0 requeues 1) 
 backlog 0b 0p requeues 1 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent :1 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 454767957 bytes 649905 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
END eth0
WORKTIME 0
BEGIN extern
qdisc htb 1: root refcnt 2 r2q 3 default 8000 direct_packets_stat 0 direct_qlen 1000
 Sent 136157 bytes 743 pkt (dropped 0, overlimits 5 requeues 0) 
 backlog 0b 0p requeues 0 
qdisc fq_codel 12: parent 1:12 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 0 bytes 0 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 11: parent 1:11 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 9752 bytes 16 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 1113 drop_overlimit 0 new_flow_count 1 ecn_mark 0
  new_flows_len 0 old_flows_len 1
qdisc fq_codel 13: parent 1:8000 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 126405 bytes 727 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 1696 drop_overlimit 0 new_flow_count 199 ecn_mark 0
  new_flows_len 0 old_flows_len 1
qdisc ingress ffff: parent ffff:fff1 ---------------- 
 Sent 268827 bytes 725 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
END extern
BEGIN extern-ifb
qdisc htb 1: root refcnt 2 r2q 21 default 8000 direct_packets_stat 0 direct_qlen 32
 Sent 341002 bytes 725 pkt (dropped 0, overlimits 1 requeues 0) 
 backlog 0b 0p requeues 0 
qdisc fq_codel 12: parent 1:12 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 0 bytes 0 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 11: parent 1:11 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 5512 bytes 8 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 689 drop_overlimit 0 new_flow_count 1 ecn_mark 0
  new_flows_len 0 old_flows_len 1
qdisc fq_codel 13: parent 1:8000 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 335490 bytes 717 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 1696 drop_overlimit 0 new_flow_count 192 ecn_mark 0
  new_flows_len 0 old_flows_len 1
END extern-ifb
BEGIN eth0
qdisc mq 0: root 
 Sent 1698194438 bytes 2677670 pkt (dropped 0, overlimits 0 requeues 6) 
 backlog 0b 0p requeues 6 
qdisc fq_codel 0: parent :8 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 2794653 bytes 9577 pkt (dropped 0, overlimits 0 requeues 1) 
 backlog 0b 0p requeues 1 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent :7 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 310537689 bytes 501842 pkt (dropped 0, overlimits 0 requeues 2) 
 backlog 0b 0p requeues 2 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent :6 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 720109 bytes 5567 pkt (dropped 0, overlimits 0 requeues 1) 
 backlog 0b 0p requeues 1 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent :5 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 559207641 bytes 940733 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent :4 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 1225494 bytes 6520 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent :3 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 367688508 bytes 556244 pkt (dropped 0, overlimits 0 requeues 1) 
 backlog 0b 0p requeues 1 
  maxpacket 387 drop_overlimit 0 new_flow_count 1 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent :2 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 1243187 bytes 7261 pkt (dropped 0, overlimits 0 requeues 1) 
 backlog 0b 0p requeues 1 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent :1 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
 Sent 454777157 bytes 649926 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0 
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
END eth0
WORKTIME 0
```

Section from netdata API re: `tc.extern_ifb`:
```
"tc.extern-ifb": 		{
			"id": "tc.extern-ifb",
			"name": "tc.extern_ifb",
			"type": "tc",
			"family": "extern-ifb",
			"context": "tc.qos",
			"title": "Class Usage (tc.extern_ifb)",
			"priority": 7000,
			"plugin": "tc",
			"module": "",
			"enabled": true,
			"units": "kilobits/s",
			"data_url": "/api/v1/data?chart=tc.extern_ifb",
			"chart_type": "stacked",
			"duration": 3996,
			"first_entry": 1520280465,
			"last_entry": 1520284461,
			"update_every": 1,
			"dimensions": {
				"15:": { "name": "15:" },
				"19:": { "name": "19:" },
				"17:": { "name": "17:" },
				"14:": { "name": "14:" },
				"12:": { "name": "12:" },
				"16:": { "name": "16:" },
				"18:": { "name": "18:" },
				"11:": { "name": "11:" },
				"13:": { "name": "13:" }
			},
			"green": null,
			"red": null,
			"alarms": {

			}
```

Section from netdata API re: `tc.extern`:
```
"tc.extern": 		{
			"id": "tc.extern",
			"name": "tc.extern",
			"type": "tc",
			"family": "extern",
			"context": "tc.qos",
			"title": "Class Usage (tc.extern)",
			"priority": 7000,
			"plugin": "tc",
			"module": "",
			"enabled": true,
			"units": "kilobits/s",
			"data_url": "/api/v1/data?chart=tc.extern",
			"chart_type": "stacked",
			"duration": 3996,
			"first_entry": 1520280465,
			"last_entry": 1520284461,
			"update_every": 1,
			"dimensions": {
				"15:": { "name": "15:" },
				"19:": { "name": "19:" },
				"17:": { "name": "17:" },
				"14:": { "name": "14:" },
				"12:": { "name": "12:" },
				"16:": { "name": "16:" },
				"18:": { "name": "18:" },
				"11:": { "name": "11:" },
				"13:": { "name": "13:" }
			},
			"green": null,
			"red": null,
			"alarms": {

			}
```

**Comment by ktsaou** [comment]
hm... spooky... although I am sure there will be a logical explanation...

Is this the latest version of netdata?

**Comment by ktsaou** [comment]
If I give you a little modified version of netdata to add extra debugging / logging, can you install it?

**Comment by seii** [comment]
I'm using version `1.9.0-1` of netdata from the [Arch Linux ARM](https://archlinuxarm.org/packages/aarch64/netdata) repository. This is the latest for my distribution, at any rate.

As for a modified version, I'd be happy to try if you gave me an idea how it would need to be installed. Would I just be replacing the binary or would further steps be needed?

**Comment by seii** [comment]
I had completely forgotten to circle back on the secondary issue we had discussed - namely, netdata not showing class names. While I don't know why, exactly, I was able to find out that my system never had the file `/etc/iproute2/tc_cls`. As soon as I created it and put in the right class names, netdata was able to show everything just fine. I would guess this is something specific to Arch Linux, although it may happen on other distros as well.
